File: web/src/services/teamService.ts
import { supabase } from '../lib/supabase'

export interface Team {
  id: string
  tournament_id: string
  name: string
  captain_id: string | null
  created_by: string
  created_at: string
}

const TEAM_FIELDS = 'id, tournament_id, name, captain_id, created_by, created_at'

export const teamService = {
  async getTeams(tournamentId: string): Promise<Team[]> {
    const { data, error } = await supabase
      .from('teams')
      .select(TEAM_FIELDS)
      .eq('tournament_id', tournamentId)
      .order('created_at')
    if (error) throw error
    return data ?? []
  },

  async getTeam(id: string): Promise<Team> {
    const { data, error } = await supabase
      .from('teams')
      .select(TEAM_FIELDS)
      .eq('id', id)
      .single()
    if (error) throw error
    return data
  },

  async createTeam(tournamentId: string, name: string): Promise<Team> {
    const { data, error } = await supabase
      .from('teams')
      .insert({ tournament_id: tournamentId, name })
      .select(TEAM_FIELDS)
      .single()
    if (error) throw error
    return data
  },
}

File: ios/Sources/SportsTracker/Team/TeamService.swift
import Foundation
import Supabase

public struct Team: Codable, Identifiable, Sendable {
    public let id: String
    public let tournamentId: String
    public let name: String
    public var captainId: String?
    public let createdBy: String
    public let createdAt: String

    enum CodingKeys: String, CodingKey {
        case id, name
        case tournamentId = "tournament_id"
        case captainId = "captain_id"
        case createdBy = "created_by"
        case createdAt = "created_at"
    }

    public init(id: String, tournamentId: String, name: String,
                captainId: String? = nil, createdBy: String = "", createdAt: String = "") {
        self.id = id
        self.tournamentId = tournamentId
        self.name = name
        self.captainId = captainId
        self.createdBy = createdBy
        self.createdAt = createdAt
    }
}

public protocol TeamServiceProtocol: Sendable {
    func getTeams(tournamentId: String) async throws -> [Team]
    func getTeam(id: String) async throws -> Team
    func createTeam(tournamentId: String, name: String) async throws -> Team
}

private let teamFields = "id, tournament_id, name, captain_id, created_by, created_at"

public final class TeamService: TeamServiceProtocol {
    private let client: SupabaseClient

    public init(client: SupabaseClient) {
        self.client = client
    }

    public func getTeams(tournamentId: String) async throws -> [Team] {
        try await client.from("teams")
            .select(teamFields)
            .eq("tournament_id", value: tournamentId)
            .order("created_at")
            .execute()
            .value
    }

    public func getTeam(id: String) async throws -> Team {
        try await client.from("teams")
            .select(teamFields)
            .eq("id", value: id)
            .single()
            .execute()
            .value
    }

    public func createTeam(tournamentId: String, name: String) async throws -> Team {
        struct Insert: Encodable {
            let tournament_id: String
            let name: String
        }
        return try await client.from("teams")
            .insert(Insert(tournament_id: tournamentId, name: name))
            .select(teamFields)
            .single()
            .execute()
            .value
    }
}
